# Lab #1

The purpose of this laboratory is to get you acquainted with Python. 
More specifically, you will learn how to:
- do some basic operations (e.g. with strings).
- work with lists and dictionaries.
- read a dataset stored as a CSV file.
- extract some useful information (mean and standard deviation) from the datasets while only using basic python.
- create a simple rule-based classifier that is already capable to perform some classification.


## Warm-up
### 0.1 Is Python installed on your system?

Check your Python installation and contextually print the *version* of you Python installation (should be in the form `3.x.x.`).

*Hint*: use `sys.version` or `os.system(...)` or a terminal command starting with `!`

In [5]:
import sys as s
print(s.version)

3.14.7 (main, Aug 14 2026, 06:38:32) [GCC 16.2.1 20260810]


### 0.2 Area of a circle **providing** the radius

Write a python script computing the area of a circle. 

The radius of the circle is provided by the user.
To provide the radius simply use the function: 
```
    radius = input(message)
```

Where message represents the name displayed before the input cell.

In [ ]:
from math import pi  # you may need pi for the computation


### Working with strings

#### 0.3 Swap two words in a string
Write a python script that and swap the word "python" with "java" and viceversa in a given string

#### 0.4 Print file extension
Write a python script that output the name and the extension of a file provided as input (as a string).

Note: the file doesn't have to *actually* exists

##

### Work with lists

##### 0.5 Revert elements order
Return a given tuple with elements in the reverse order
e.g. given the tuple `(2, 4, 5, 8)` the script should return `(8, 5, 4, 2)`


#### 0.6 Euclidean distance between two vectors
Compute the euclidean distances between two vectors (passed as lists) as following:

$ dist(v1, v2) = \sqrt{\sum_i{(v1_i-v2_i)^2}} $

*hint*: you can use the `math.sqrt` function to compute the square root of a number

*Can you do it in just one line?*

#

### Work with dictionaries

##### 0.7 Swap keys and values
Write a script reading a dictionary and returning a new dictionary with the keys and values swapped.
e.g. given the dictionary `{"a": 1, "b": 2, "c": 3}` the script should return `{1: "a", 2: "b", 3: "c"}`


#### 0.8 Dictionary concatenations
Write a script to concatenate the following dictionaries to create a new one.

*Note*: you can use **at most** a single loop

e.g.
```
dict1={1:10, 2:20}
dict2={3:30, 4:40}
dict3={5:50, 6:60}
```

Expected Result: `{1: 10, 2: 20, 3: 30, 4: 40, 5: 50, 6: 60}`

#

#

## Iris analysis and first classification exercise
### Dataset Download

### **Iris**
Iris is a particularly famous *toy dataset* (i.e. a dataset with a small number of rows and columns, mostly
used for initial small-scale tests and proofs of concept). 
This specific dataset contains information about the **Iris**, a genus that includes 260-300 species of plants. 
The Iris dataset contains measurements for 150 Iris flowers, each belonging to one of three species (50 flowers each): 

Iris Virginica             |  Iris Versicolor          |   Iris Setosa  |
:-------------------------:|:-------------------------:|:---------------|
<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/f/f8/Iris_virginica_2.jpg/1280px-Iris_virginica_2.jpg" alt="Iris Virginica" width="200" by="By Eric Hunt - Own work, CC BY-SA 4.0, https://commons.wikimedia.org/w/index.php?curid=72555262" /> | <img src="https://upload.wikimedia.org/wikipedia/commons/thumb/2/27/Blue_Flag%2C_Ottawa.jpg/1280px-Blue_Flag%2C_Ottawa.jpg" alt="Iris Versicolor" width="200" by="By D. Gordon E. Robertson - Own work, CC BY-SA 3.0, https://commons.wikimedia.org/w/index.php?curid=10227368" /> |<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/86/Iris_setosa.JPG/1280px-Iris_setosa.JPG" alt="Iris Setosa" width="200" by="By Qwert1234 - Qwert1234's file, Public Domain, https://commons.wikimedia.org/w/index.php?curid=9060295" />|

Each of the 150 flowers contained in the Iris dataset is represented by 5 values:
- sepal length, in cm
- sepal width, in cm
- petal length, in cm
- petal width, in cm
- Iris species, one of: Iris-setosa, Iris-versicolor, Iris-virginica (the label)

Each row of the dataset represents a distinct flower (as such, the dataset will have 150 rows). Each
row then contains 5 values (4 measurements and a species label).
The dataset is described in more detail on the [UCI Machine Learning Repository website](https://archive.ics.uci.edu/dataset/53/iris). The dataset
can either be downloaded directly from there (iris.data file), or from a terminal, using the `wget` tool (linux users) or using the corresponding `wget` python library (windows users). The
following command downloads the dataset from the original URL and stores it in a file named iris.csv.

```
# linux users
! wget "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data" -O iris.csv
# windows users
! pip install wget
import wget
wget.download("https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data", "iris.csv")
```

The dataset is available as a Comma-Separated Values (CSV) file. These files are typically used to
represent tabular data. 
- Each row is represented on one of the lines. 
- Each of the rows contains a fixed number of columns. 
- Each of the columns (in each row) is separated by a comma (,).

To read CSV files, Python offers a module called `csv` (here the offical [doc](https://docs.python.org/3/library/csv.html)). This module allows using `csv.reader()`, which
reads a file row by row. For each row, it returns a list of columns that can be processed as needed. 


Let's download the dataset and print the first three rows.




In [7]:
# linux users
! wget "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data" -O iris.csv
# windows users
# pip install wget

--2026-10-02 16:32:55--  https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data
Loaded CA certificate '/etc/ssl/certs/ca-certificates.crt'
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: unspecified
Saving to: ‘iris.csv’

iris.csv                [ <=>                ]   4,44K  --.-KB/s    in 0s      

2026-10-02 16:32:56 (17,8 MB/s) - ‘iris.csv’ saved [4551]



In [8]:
# import wget
# wget.download("https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data", "iris.csv") # download in the current directory

print("Reading first lines of IRIS dataset")
import csv 
with open("iris.csv") as f:
    for i, rows in enumerate(csv.reader(f)):
        print(rows)
        if (i+1) >= 3:
            break

Reading first lines of IRIS dataset
['5.1', '3.5', '1.4', '0.2', 'Iris-setosa']
['4.9', '3.0', '1.4', '0.2', 'Iris-setosa']
['4.7', '3.2', '1.3', '0.2', 'Iris-setosa']


Note by default, csv.reader converts all fields read into strings (str). 
If you want to treat them as number, remember to **cast them** correctly!

## Exercises
Note that exercises marked with a (*) are optional, you should focus on completing the other ones first.
### Iris analysis
1. Load the previously downloaded Iris dataset as a list of lists (each of the 150 lists should have 5 elements). You can make use of the csv module presented

In [ ]:
filename = "iris.csv"
flowers = []
with open(filename, mode='r') as f:
    for row in csv.reader(f):
        if len(row) > 0:
            flowers.append(row)

print("Number of flowers in the dataset: ", len(flowers))
print(f"First flower: {flowers[0]}")
print(f"Number of attributes for each flower: {len(flowers[0])}")


Number of flowers in the dataset:  150
First flower: ['5.1', '3.5', '1.4', '0.2', 'Iris-setosa']
Number of attributes for each flower: 5


2. Compute and print the mean and the standard deviation for each of the 4 measurement columns (i.e. sepal length and width, petal length and width). Remember that, for a given list of n values $x = (x_1, x_2, ..., x_n)$, the mean $\mu$ and the standard deviation $\sigma$ are defined respectively as:
$$\mu = {1 \over n} \sum_i^n x_i $$

$$ \sigma = \sqrt{ {1 \over n} \sum_i^n (x_i - \mu)^2} $$

In [19]:
import math

def mean(values):
    somma = 0
    for v in values:
        somma = somma + float(v)
    return somma / len(values)

def standard_deviation(values):
    m = mean(values)
    somma_quadrati = 0
    for v in values:
        scarto = float(v) - m
        somma_quadrati = somma_quadrati + scarto ** 2
    varianza = somma_quadrati / len(values)
    return math.sqrt(varianza)


means = []
stds = []

for i in range(4):
    # estraggo la colonna i
    colonna = []
    for row in flowers:
        colonna.append(row[i])

    # applico le nostre funzioni alla colonna
    means.append(mean(colonna))
    stds.append(standard_deviation(colonna))

print("Means", means)
print("Standard deviations", stds)

Means [5.843333333333335, 3.0540000000000007, 3.7586666666666693, 1.1986666666666672]
Standard deviations [0.8253012917851409, 0.4321465800705435, 1.7585291834055201, 0.760612618588172]



3. Compute and print the mean and the standard deviation for each of the 4 measurement columns, separately for each of the three Iris species (versicolor, virginica and setosa).

*Hint*: use the `filter()` function with a lambda function to keep only the rows that `row[4] == <one of the iris species>`

In [ ]:
# --- Estrazione delle specie ---
species = set()
for row in flowers:
    species.add(row[4])

species = sorted(species)
print(f"Species: {species}")


# --- Raggruppamento dei fiori per specie ---
groups = {}
for row in flowers:
    specie = row[4]
    if specie not in groups:
        groups[specie] = []        # prima volta che vedo questa specie: creo la lista vuota
    groups[specie].append(row)     # aggiungo il fiore alla lista della sua specie

for specie in species:
    print(f"{specie}: {len(groups[specie])} flowers")


# --- Media e deviazione standard per ogni specie ---
for specie in species:
    subset = groups[specie]

    means = []
    stds = []
    for i in range(4):
        colonna = []
        for row in subset:
            colonna.append(row[i])
        means.append(mean(colonna))
        stds.append(standard_deviation(colonna))

    print(specie)
    print("  Means", means)
    print("  Standard deviations", stds)

Species: ['Iris-setosa', 'Iris-versicolor', 'Iris-virginica']
Iris-setosa: 50 flowers
Iris-versicolor: 50 flowers
Iris-virginica: 50 flowers
Iris-setosa
  Means [5.005999999999999, 3.4180000000000006, 1.464, 0.2439999999999999]
  Standard deviations [0.348946987377739, 0.37719490982779713, 0.17176728442867115, 0.10613199329137278]
Iris-versicolor
  Means [5.936, 2.7700000000000005, 4.26, 1.3259999999999998]
  Standard deviations [0.5109833656783752, 0.31064449134018135, 0.4651881339845204, 0.19576516544063702]
Iris-virginica
  Means [6.587999999999998, 2.9739999999999998, 5.552, 2.026]
  Standard deviations [0.6294886813914925, 0.319255383666431, 0.5463478745268441, 0.2718896835115301]



4. Based on the results of exercises 2 and 3, which of the 4 measurements would you considering as being the most characterizing one for the three species? (In other words, which measurement would you consider “best”, if you were to guess the Iris species based only on those four values?)

**Petal width (the 4th measurement) is the most characterizing one: its means are well separated across the three species (0.24, 1.33 and 2.03 cm) and its standard deviations are small (0.11-0.27 cm), so the values of different species overlap little. Petal length (the 3rd measurement) is almost as good (means 1.46, 4.26 and 5.55 cm, standard deviations 0.17-0.55 cm) and is an acceptable answer too.**


5. Based on the considerations of Exercise 4, assign the flowers with the following measurements to what you consider would be the most likely species.
````
5.2, 3.1, 4.0, 1.2 : versicolor
4.9, 2.5, 5.6, 2.0 : virginica
5.4, 3.2, 1.9, 0.4 : setosa
````


6. (*) Create a Rule-based classifier that will receive some rule and will classify each sample into one of the three species.  According to your analysis in the previous point where you identified the most discriminative feature, provide the classifier with 3 rules, one for classifying each iris, based on this feature.

7. (*) Compute prediction for all the elements in the dataset and store them in a list. Then, compute the accuracy of the classifier that you create. Remember that the accuracy metric is:

$$ {\text{number of correct predictions} \over \text{total number of predictions}} $$

Where one can check whether the prediction is correct by looking at the label of the sample ($5^{th}$ column)

#